src/ripple/basics/impl/UptimeTimer.cpp

```diff
@@ -19,64 +19,53 @@
 
 #include <ripple/basics/UptimeTimer.h>
 
-#include <atomic>
-
 namespace ripple {
 
-UptimeTimer::UptimeTimer ()
-    : m_elapsedTime (0)
-    , m_startTime (::time (0))
-    , m_isUpdatingManually (false)
-{
-}
+std::atomic<UptimeClock::rep>  UptimeClock::now_{0};      // seconds since start
+std::atomic<bool>              UptimeClock::stop_{false}; // stop update thread
 
-UptimeTimer::~UptimeTimer ()
+// On rippled shutdown, cancel and wait for the update thread
+UptimeClock::update_thread::~update_thread()
 {
-}
-
-int UptimeTimer::getElapsedSeconds () const
-{
-    int result;
-
-    if (m_isUpdatingManually)
-    {
-        std::atomic_thread_fence (std::memory_order_seq_cst);
-        result = m_elapsedTime;
-    }
-    else
+    if (joinable())
     {
-        // VFALCO TODO use time_t instead of int return
-        result = static_cast <int> (::time (0) - m_startTime);
+        stop_ = true;
+        join();
     }
-
-    return result;
-}
-
-void UptimeTimer::beginManualUpdates ()
-{
-    //assert (!m_isUpdatingManually);
-
-    m_isUpdatingManually = true;
 }
 
-void UptimeTimer::endManualUpdates ()
+// Launch the update thread
+UptimeClock::update_thread
+UptimeClock::start_clock()
 {
-    //assert (m_isUpdatingManually);
-
-    m_isUpdatingManually = false;
+    return update_thread{[]
+                         {
+                             using namespace std;
+                             using namespace std::chrono;
+
+                             // Wake up every second and update now_
+                             auto next = system_clock::now() + 1s;
+                             while (!stop_)
+                             {
+                                 this_thread::sleep_until(next);
+                                 next += 1s;
```

Comment:
If the thread sleeps longer than `next` but by less than 1s, then `now_` is indeed incremented late (which would also happen if computed), but the next time through the loop the thread will sleep for less than a second, proportional to the amount it was late.  This is because `next` is _always_ incremented by exactly 1s and never impacted by lateness.

And if the sleep is late by 1s or more, same thing except now next time through the loop there is no sleeping at all, so `now_` would increment rapidly until it is caught up to accurate.

The previous implementation lost time under "manual increment" because it used `sleep_for` instead of `sleep_until`.  `sleep_for` can't take into account time spent outside the sleep in order to keep a regular time.  The previous implementation did attempt to correct for this, but it is impossible to do it perfectly with `sleep_for`.

Reply:
Thanks for the explanation.  Excellent point on the distinction between `sleep_for` and `sleep_until` which I had entirely missed.